VBF_Cat update

valeriadamante opened the pull request:
Changed files:
- config/2018/sources.cfg --> added the option mdnn_version and added in draw option data
- include/AnalysisCategories.h --> added the VBF_Category enum, their names, substitution of VBF_wp with vbf_cat
- include/EventTags.h --> added the declaration of the method FindVBFCategory
- source/ProcessAnaTuple.cxx --> added in the TupleReader the mdnn_version
- src/AnalysisCategories.cpp --> added the map for the VBF categories names, change in Contains function
- src/EventTags.cpp --> Added the method FindVBFCategory and other changes in the CreateEventTags method
- src/AnaTupleReader.cpp --> Added the VBF Category and mdnn_score branches
- include/AnaTupleReader.h --> added the mdnn_version in DefineBranches constructor

kandrosov commented on Analysis/src/EventTags.cpp:
```diff
@@ -19,24 +19,35 @@ EventTagCreator::EventTagCreator(const EventCategorySet& _categories,
-
-    return vbf_tag_raw;
+    else category_and_dnn.push_back(std::make_pair(-1.f, VBF_Categories::None));
+    std::sort(category_and_dnn.begin(), category_and_dnn.end());
```
with `std::max_element` is better ;)
```c++
return *std::max_element(category_and_dnn.begin(), category_and_dnn.end());
```

kandrosov commented on Analysis/src/EventTags.cpp:
```diff
@@ -48,49 +59,45 @@ EventTags EventTagCreator::CreateEventTags(const std::vector<DataId>& dataIds_ba
-    if(vbf_tag_raw > 0) vbf_tag = static_cast<DiscriminatorWP>(vbf_tag_raw);
+    std::map<DiscriminatorWP, float > btag_weights = {};
+    if(btag_weight_Loose.size()!=0 || btag_weight_Medium.size()!=0 || btag_weight_Tight.size()!=0) btag_weights = {
+      { DiscriminatorWP::Loose, btag_weight_Loose.at(0) },
```
- define map as `std::map<DiscriminatorWP, const std::vector<float>* >`;
- create a lambda to get weight given unc and vector as arguments (like this it will be universal for all weights and unc variations).

valeriadamante replied:
Which type is used to pass the unc?

kandrosov replied:
`UncertaintyScale` which is defined [here](https://github.com/hh-italian-group/h-tautau/blob/master/Core/include/AnalysisTypes.h#L38)

The definition of how up/down weights are stored in anaTuple is defined [here](https://github.com/hh-italian-group/hh-bbtautau/blob/master/Analysis/src/AnaTuple.cpp#L96-L111). The new lambda should basically perform an inverse operation and return the weight for the given scale.

kandrosov commented on Analysis/src/EventTags.cpp:
```diff
@@ -48,49 +59,45 @@ EventTags EventTagCreator::CreateEventTags(const std::vector<DataId>& dataIds_ba
-                }
-                if(use_kinFit)
+            if(has_b_pair) {
+              const EllipseParameters& window = category.HasBoostConstraint() && category.IsBoosted() ? cuts::hh_bbtautau_Run2::hh_tag::boosted_window : cuts::hh_bbtautau_Run2::hh_tag::resolved_window;
```
Have you changed definitions of mass windows in the header? If yes, you also need to do a PR in h-tautau repository.

kandrosov commented on Analysis/src/EventTags.cpp:
```diff
@@ -48,49 +59,45 @@ EventTags EventTagCreator::CreateEventTags(const std::vector<DataId>& dataIds_ba
-                if(use_kinFit)
+            if(has_b_pair) {
+              const EllipseParameters& window = category.HasBoostConstraint() && category.IsBoosted() ? cuts::hh_bbtautau_Run2::hh_tag::boosted_window : cuts::hh_bbtautau_Run2::hh_tag::resolved_window;
+              if(use_svFit){
```
remove `use_svFit` and `use_kinFit` as arguments from everywhere in ProcessAnaTuple and EventTag.

kandrosov commented on Analysis/src/EventTags.cpp:
```diff
@@ -48,49 +59,45 @@ EventTags EventTagCreator::CreateEventTags(const std::vector<DataId>& dataIds_ba
+            if(has_b_pair) {
+              const EllipseParameters& window = category.HasBoostConstraint() && category.IsBoosted() ? cuts::hh_bbtautau_Run2::hh_tag::boosted_window : cuts::hh_bbtautau_Run2::hh_tag::resolved_window;
+              if(use_svFit){
+                evtSubCategory.SetCutResult(SelectionCut::mh, window.IsInside(SVfit_p4.M(), m_bb));
```
you need to require that SVfit has a valid value.
```c++
evtSubCategory.SetCutResult(SelectionCut::mh, SVfit_valid && window.IsInside(SVfit_p4.M(), m_bb));
```

kandrosov commented on Analysis/src/EventTags.cpp:
```diff
@@ -48,49 +59,45 @@ EventTags EventTagCreator::CreateEventTags(const std::vector<DataId>& dataIds_ba
+              }
+              if(use_kinFit)
                     evtSubCategory.SetCutResult(SelectionCut::KinematicFitConverged,
                                                 kinFit_convergence);
```
`kinFit_convergence > 0`

valeriadamante commented on Analysis/include/EventTags.h:
```diff
@@ -26,12 +26,19 @@ class EventTagCreator {
+    std::pair<float, VBF_Categories> FindVBFCategory(float dnn_score_TT_dl, float dnn_score_DY, float dnn_score_TT_lep,
+                                                     float dnn_score_qqHH_sm, float dnn_score_ggHH, float dnn_score_TT_FH,
+                                                     float dnn_score_ttH, float dnn_score_ttH_tautau, float dnn_score_tth_bb,
+                                                     float dnn_score_qqHH, float dnn_score_qqHH_vbf_c2v, float dnn_score_TT_sl, std::string mdnn_version) const;
```
I introduced here the ```if``` inside the FindVBFCategory method so that, in Anatuple.cpp, the branch "vbf_cat" is defined out of the if-else

kandrosov replied:
in this way, the check will be done for each event, which is not necessary. I suggest removing mdnn_version from arguments.

valeriadamante commented on Analysis/src/AnaTuple.cpp:
```diff
@@ -482,27 +484,41 @@ void AnaTupleReader::DefineBranches(const NameSet& active_var_names, bool all, c
+                                    float dnn_score_qqHH, float dnn_score_qqHH_vbf_c2v, float dnn_score_qqHH_sm){
+                                    return event_tagger.FindVBFCategory( dnn_score_TT_dl,  dnn_score_TT_sl,  dnn_score_TT_lep, dnn_score_TT_FH,  dnn_score_DY,  dnn_score_ggHH, dnn_score_ttH,  dnn_score_ttH_tautau,  dnn_score_tth_bb, dnn_score_qqHH,  dnn_score_qqHH_vbf_c2v,  dnn_score_qqHH_sm, mdnn_version);
+                                };
+    Define(df, "vbf_cat", vbf_category, mdnn_branches, true);
```
is the dummy method still required if I do like the following?
``` 
const auto vbf_category = [&](float dnn_score_TT_dl, float dnn_score_TT_sl, float dnn_score_TT_lep,
                                    float dnn_score_TT_FH, float dnn_score_DY, float dnn_score_ggHH,
                                    float dnn_score_ttH, float dnn_score_ttH_tautau, float dnn_score_tth_bb,
                                    float dnn_score_qqHH, float dnn_score_qqHH_vbf_c2v, float dnn_score_qqHH_sm){
                                            return event_tagger.FindVBFCategory( dnn_score_TT_dl,  dnn_score_TT_sl,  dnn_score_TT_lep, dnn_score_TT_FH,
                                            dnn_score_DY,  dnn_score_ggHH, dnn_score_ttH,  dnn_score_ttH_tautau,  dnn_score_tth_bb, dnn_score_qqHH,  dnn_score_qqHH_vbf_c2v,  dnn_score_qqHH_sm);
                                };
                            }
    if(mdnn_version.empty()){
        auto vbf_cat = [](){return std::make_pair(-1.f, VBF_Categories::None);};
        Define(df, "vbf_cat", vbf_cat, {}, true);
    }
    else{
        Define(df, "vbf_cat", vbf_category, mdnn_branches, true);
        Define(df, "mdnn_score", [](const std::pair<float, VBF_Categories>& vbf_cat) -> float { return vbf_cat.first; }, {"vbf_cat"}, true);
    }
```

kandrosov replied:
`auto vbf_cat = ...` is the dummy method I had in mind, so, should be ok as you have written. With small correction that `mdnn_score` definition should be done outside if/else.

kandrosov commented on Analysis/include/AnalysisCategories.h:
```diff
@@ -133,13 +128,13 @@ struct EventCategory {
     boost::optional<DiscriminatorWP> btag_wp;
     boost::optional<bool> boosted, is_VBF;
-    VBF_Categories vbf_cat;
+    VBF_Category vbf_cat(VBF_Category::None);
```
```c++
VBF_Category vbf_cat{VBF_Category::None};
```

kandrosov commented on Analysis/include/AnalysisCategories.h:
```diff
@@ -238,3 +233,4 @@ using EventSubCategorySet = std::set<EventSubCategory>;
 using Dataset = std::string;
 
 } // namespace analysis
+VBF_Category
```
what is this line?

kandrosov commented on Analysis/include/EventTags.h:
```diff
@@ -23,10 +23,9 @@ class EventTagCreator {
+                    const std::set<UncertaintySource>& _norm_unc_sources);
 
-    std::pair<float, VBF_Categories> FindVBFCategory(float dnn_score_TT_dl, float dnn_score_DY, float dnn_score_TT_lep,
+    static const std::pair<float, VBF_Category>& FindVBFCategory(float dnn_score_TT_dl, float dnn_score_DY, float dnn_score_TT_lep,
```
Never return a reference or a pointer to a temporary variable!
```c++
static std::pair<float, VBF_Category> FindVBFCategory(...);
```

kandrosov commented on Analysis/src/AnaTuple.cpp:
```diff
@@ -494,19 +494,26 @@ void AnaTupleReader::DefineBranches(const NameSet& active_var_names, bool all, c
+    }
+    else{
+        Define(df, "vbf_cat", vbf_category, mdnn_branches, true);
+        Define(df, "mdnn_score", [](const std::pair<float, VBF_Categories>& vbf_cat) -> float { return vbf_cat.first; }, {"vbf_cat"}, true);
```
`mdnn_score` should be defined outside of the loop

kandrosov commented on Analysis/src/AnalysisCategories.cpp:
```diff
@@ -315,8 +315,8 @@ std::string EventCategory::ToString() const
-            if(vbf_cat != VBF_Categories::None)
-                VBF_str += "_" + VBF_Categories_names.EnumToString(vbf_cat);
+            if(vbf_cat != VBF_Category::None)
+                VBF_str += "_" + VBF_Category.analysis::ToString(vbf_cat);
```
```c++
VBF_str += "_" + ::analysis::ToString(vbf_cat);
```

kandrosov commented on Analysis/src/AnalysisCategories.cpp:
```diff
@@ -328,15 +328,15 @@ EventCategory EventCategory::Parse(const std::string& str)
-          { "VBF_"+ VBF_Categories_names.EnumToString(VBF_Categories::TT_FH),  { true, VBF_Categories::TT_FH }},
-          { "VBF_"+ VBF_Categories_names.EnumToString(VBF_Categories::ttH),  { true, VBF_Categories::ttH }},
-          { "VBF_"+ VBF_Categories_names.EnumToString(VBF_Categories::DY),  { true, VBF_Categories::DY }} };
+    static const std::map<std::string, std::pair<bool, VBF_Category>> VBF_suffix =
```
eliminate `VBF_suffix` map

kandrosov reviewed (changes requested):
undo changes for file permissions

kandrosov commented on Analysis/config/2018/sources.cfg:
```diff
@@ -3,22 +3,22 @@ int_lumi: 59741
 #categories: 2j 2j0bR_noVBF 2j1bR_noVBF 2j2b+R_noVBF 2j2Lb+B_noVBF 2j1b+_VBF 2j1b+_VBFT 2j1b+_VBFL 2j0b 2j1b 2j2b+ 2j0Lb 2j1Lb 2j2Lb+ 2j0Tb 2j1Tb 2j1Tb+ 2j2Tb+
-categories: 2j 2j1bR_noVBF 2j2b+R_noVBF 2j2Lb+B_noVBF 2j1b+_VBF
-sub_categories: mh
+categories: 2j 2j1bR_noVBF 2j2b+R_noVBF 2j2Lb+B_noVBF 2j1b+_VBF 2j1b+_VBF_qqHH 2j1b+_VBF_ggHH 2j1b+_VBF_TT_L 2j1b+_VBF_TT_FH 2j1b+_VBF_ttH 2j1b+_VBF_DY
+sub_categories: mh NoCuts
```
Apply similar changes for 2016&2017

kandrosov commented on Analysis/config/histograms.cfg:
```diff
@@ -166,6 +170,13 @@ mva_score/tauTau/2j1b+_VBFT: x_bins="0 0.025 0.05 0.15 0.35 0.55 0.7 0.85 0.95 0
                blind_ranges=0.7:1
+dnn_score_kl1/*/2j: x_range=0:1/20 x_title="DNN score" y_title="dN / bin width" log_y=true max_y_sf=100 \
+
+mdnn_score: x_range=0:1/20 x_title="DNN score" y_title="dN / bin width" log_y=true max_y_sf=100 \
+               blind_ranges=0.7:1
+mdnn_score/*/2j: x_range=0:1/20 x_title="DNN score" y_title="dN / bin width" log_y=true max_y_sf=100 \
```
`x_title="MDNN score"`

kandrosov commented on Analysis/include/EventTags.h:
```diff
@@ -21,25 +21,36 @@ class EventTagCreator {
+                     const std::set<UncertaintySource>& _norm_unc_sources,
+                     bool _use_kinFit, bool _use_svFit);
 
     static int CreateVBFTag(const LorentzVectorM& VBF1_p4, const LorentzVectorM& VBF2_p4, bool is_VBF,
                      bool pass_vbf_trigger);
+    //int CreateVBFTag(const LorentzVectorM& VBF1_p4, const LorentzVectorM& VBF2_p4, bool is_VBF,
+    //                 bool pass_vbf_trigger) const;
```
CreateVBFTag can be removed

kandrosov commented on Analysis/include/EventTags.h:
```diff
@@ -21,25 +21,36 @@ class EventTagCreator {
+                     const std::map<SelectionCut, analysis::EllipseParameters>& _massWindowParams,
+                     const std::set<UncertaintySource>& _event_unc_sources,
+                     const std::set<UncertaintySource>& _norm_unc_sources,
+                     bool _use_kinFit, bool _use_svFit);
```
- Add flag whatever to use iterative fit btag SFs or combined ones.
- remove use_kinFit and use_svFit

kandrosov commented on Analysis/include/EventTags.h:
```diff
@@ -21,25 +21,36 @@ class EventTagCreator {
-                              float m_bb, float m_tt_vis, int kinFit_convergence) const;
+                               const std::vector<float>& btag_weight_Loose,
+                               const std::vector<float>& btag_weight_Medium,
+                               const std::vector<float>& btag_weight_Tight,
```
also btag_weights_iterativeFit

kandrosov commented on Analysis/include/EventTags.h:
```diff
@@ -21,25 +21,36 @@ class EventTagCreator {
+                               const std::vector<float>& btag_weight_Medium,
+                               const std::vector<float>& btag_weight_Tight,
+                               int num_central_jets, bool has_b_pair, int num_btag_loose, int num_btag_medium,
+                               int num_btag_tight, bool is_vbf, bool is_boosted, std::pair<float, VBF_Category>& vbf_cat,
```
`const std::pair<float, VBF_Category>&`

kandrosov commented on Analysis/include/EventTags.h:
```diff
@@ -21,25 +21,36 @@ class EventTagCreator {
     const EventSubCategorySet& subCategories;
     const std::map<SelectionCut,analysis::EllipseParameters>& massWindowParams;
     const std::set<UncertaintySource>& event_unc_sources, norm_unc_sources;
+    //const bool use_kinFit, use_svFit;
     const bool use_kinFit, use_svFit;
```
remove use_kinFit and use_svFit

kandrosov commented on Analysis/src/AnaTupleReader.cpp:
```diff
@@ -46,13 +46,12 @@ std::vector<std::shared_ptr<TTree>> AnaTupleReader::ReadTrees(Channel channel,
 
 AnaTupleReader::AnaTupleReader(const std::string& file_name, Channel channel, NameSet& active_var_names,
-                               const std::vector<std::string>& input_friends, const EventTagCreator& event_tagger) :
-        files(OpenFiles(file_name, input_friends)), trees(ReadTrees(channel, files)), dataFrame(*trees.front()), df(dataFrame)
```
why this line is removed?

kandrosov commented on Analysis/src/AnaTupleReader.cpp:
```diff
@@ -194,10 +193,27 @@ void AnaTupleReader::DefineBranches(const NameSet& active_var_names, bool all, c
     Define(df, "event_tags", create_event_tags,
-        { "dataIds_base", "all_weights", "num_central_jets", "has_b_pair", "num_btag_Loose", "num_btag_Medium",
-          "num_btag_Tight", "is_vbf", "is_boosted", "vbf_tag_raw", "SVfit_p4", "MET_p4", "m_bb", "m_tt_vis",
+        { "dataIds_base", "all_weights", "btag_weight_IterativeFit", "btag_weight_IterativeFit","btag_weight_IterativeFit","num_central_jets", "has_b_pair", "num_btag_Loose", "num_btag_Medium",
```
pass all 4 btag_weight branches

kandrosov commented on Analysis/src/AnalysisCategories.cpp:
```diff
@@ -269,11 +267,17 @@ bool EventCategory::isVBF() const
+    if(!HasVBFCategory())
+        throw exception("VBF category is not defined.");
+    return *vbf_cat;
+}*/
```
remove commented part

valeriadamante commented:
in changed files:
- added the mdnn_version, subcategory 'noCuts' and VBF categories in all sources.cfg
- modified the mdnn histogram: x-axis name "dnn_score" -> "mdnn_score"
- removed createVBFtag from all files
- removed the flags use_SVFit and use_Kinfit  (so all the files containing the svfit and kinfit flags have been changed)
- added the btag_weight_iterativeFit and the corresponding flags

kandrosov commented on Analysis/src/EventTags.cpp:
```diff
@@ -90,9 +116,8 @@ EventTags EventTagCreator::CreateEventTags(const std::vector<DataId>& dataIds_ba
-                if(use_kinFit)
-                    evtSubCategory.SetCutResult(SelectionCut::KinematicFitConverged,
-                                                kinFit_convergence);
+                */
```
remove commented part
